kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/DomainMetadataUtils.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (2024) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.internal.util;
+
+import io.delta.kernel.data.Row;
+import io.delta.kernel.internal.DeltaErrors;
+import io.delta.kernel.internal.actions.DomainMetadata;
+import io.delta.kernel.internal.actions.Protocol;
+import io.delta.kernel.types.StructType;
+import io.delta.kernel.utils.CloseableIterable;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class DomainMetadataUtils {
+
+  /**
+   * Extracts a map of domain metadata from actions. In one transaction, there can be only one
+   * DomainMetadata action per domain name. It throws a KernelException if duplicate domain metadata
+   * actions is found.
+   *
+   * @param actions an iterable of Row objects representing actions
+   * @param schema the schema of the row
+   * @return a map where the keys are domain names and the values are DomainMetadata objects
+   */
+  public static Map<String, DomainMetadata> extractDomainMetadataMap(
+      CloseableIterable<Row> actions, StructType schema) {
+    Map<String, DomainMetadata> domainMetadataMap = new HashMap<>();
+
+    final int domainMetadataOrdinal = schema.indexOf("domainMetadata");
+    if (domainMetadataOrdinal == -1) {
+      throw new IllegalArgumentException("Schema does not contain 'domainMetadata' field");
```

Comment:
@allisonport-db @vkorukanti -- when should we be using `KernelException` or not?